src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/level/MixinLevel.java

```diff
@@ -1,44 +1,209 @@
+    private void cc_replaceLevelChunkInGetBlockState(BlockPos pos, CallbackInfoReturnable<BlockState> cir, @Share("levelCube") LocalRef<LevelCube> levelCubeLocalRef) {
+        if(cc_isCubic) {
+            levelCubeLocalRef.set(this.cc_getCubeAt(pos));
+        }
+    }
+
+    @WrapOperation(method = "getBlockState", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/chunk/LevelChunk;getBlockState(Lnet/minecraft/core/BlockPos;)"
+        + "Lnet/minecraft/world/level/block/state/BlockState;"))
+    private BlockState cc_replaceLevelChunkInGetBlockState(LevelChunk levelChunk, BlockPos pos, Operation<BlockState> original,
+                                                           @Share("levelCube") LocalRef<LevelCube> levelCubeLocalRef) {
+        if(cc_isCubic) {
+            return levelCubeLocalRef.get().getBlockState(pos);
+        }
+        return original.call(levelChunk, pos);
+    }
+
+    // getBlockEntity
+    @Inject(method = "getBlockEntity", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/Level;getChunkAt(Lnet/minecraft/core/BlockPos;)"
+        + "Lnet/minecraft/world/level/chunk/LevelChunk;"), cancellable = true)
+    private void cc_replaceGetChunkAtInSetBlockEntity(BlockPos pos, CallbackInfoReturnable<BlockEntity> cir) {
+        if(cc_isCubic) {
+            cir.setReturnValue(this.cc_getCubeAt(pos).getBlockEntity(pos, LevelChunk.EntityCreationType.IMMEDIATE));
+        }
+    }
+
+    // getFluidState
+    @WrapOperation(method = "getFluidState", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/chunk/LevelChunk;getFluidState(Lnet/minecraft/core/BlockPos;)Lnet/minecraft/world/level/material/FluidState;"))
+    private FluidState cc_replaceGetChunkAtInGetFluidState(LevelChunk levelChunk, BlockPos pos, Operation<FluidState> original) {
+        if(cc_isCubic) {
+            return this.cc_getCubeAt(pos).getFluidState(pos);
+        }
+        return original.call(levelChunk, pos);
+    }
+
+    // setBlockEntity
+    @Inject(method = "setBlockEntity", at = @At(value="INVOKE", target="Lnet/minecraft/world/level/Level;getChunkAt(Lnet/minecraft/core/BlockPos;)"
+        + "Lnet/minecraft/world/level/chunk/LevelChunk;"), cancellable=true)
+    private void cc_replaceLevelChunkInSetBlockEntity(BlockEntity blockEntity, CallbackInfo ci, @Local(ordinal = 0) BlockPos pos) {
+        if(cc_isCubic) {
+            this.cc_getCubeAt(pos).addAndRegisterBlockEntity(blockEntity);
+            ci.cancel();
+        }
+    }
+
+    // removeBlockEntity
+    @Inject(method = "removeBlockEntity", at = @At(value="INVOKE", target="Lnet/minecraft/world/level/Level;getChunkAt(Lnet/minecraft/core/BlockPos;)Lnet/minecraft/world/level/chunk/LevelChunk;"), cancellable=true)
+    private void cc_replaceGetChunkAtInRemoveBlockEntity(BlockPos pos, CallbackInfo ci) {
+        if(cc_isCubic) {
+            this.cc_getCubeAt(pos).removeBlockEntity(pos);
+            ci.cancel();
```

Comment:
cancelling here prevents `updateNeighbourForOutputSignal` from running.
Perhaps this should be two redirects targetting `getChunkAt` and `removeBlockEntity`?

Reply:
This also makes me worry about the mixin above this: `cc_replaceLevelChunkInSetBlockEntity`. Currently it works fine but if vanilla adds logic below the `if`, it would break in a similar way